### DISCUSSION TITLE
IPAM deployment failed

### DISCUSSION THREAD

**Discussion by veerabose** [Azure/ipam#220] [Q&A]
Hi Experts, 

I tried to deploy and configure an Azure IPAM on a free tire subscription with full tenant ownership on a windows 10 system.  I installed "Az"; ModuleVersion="11.1.0"; and modified Az model version to 11.1.0 in deploy.ps1, I executed the "./deploy.ps1 -Location "eastus"", I got the following error; could you please guide me to complete the successful  IMPA deployment.

deploy_log
-----------
NOTE: IPAM Deployment Type: Full
INFO: Fetching Tenant ID from Azure PowerShell SDK
INFO: Fetching Azure Cloud type from Azure PowerShell SDK
ERROR: Unable to deploy Azure IPAM solution due to an exception, see logs for detailed information!
Run Log: ..\logs\deploy_20231222070723PM.log
Error Log: ..\logs\error_20231222070723PM.log

error_log
InvalidOperation: C:\Users\Admin\ipam\deploy\deploy.ps1:808
Line |
 808 |        $azureCloud = $AZURE_ENV_MAP[(Get-AzContext).Environment.Name]
     |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Index operation failed; the array index evaluated to null.

Thanks & Regards,
Veera.

**Comment by DCMattyG** [comment]
Hi @veerabose, appreciate you reaching out for support!

First and foremost, you shouldn't need to modify anything in the `deploy.ps1` script. The only check there is for the **minimum** version of PowerShell and that check is also there to prevent known issues.

I also wanted to make sure that when you upgraded the PowerShell SDK that you followed the instructions located in the following doc:

https://learn.microsoft.com/en-us/powershell/azure/install-azps-windows?view=azps-11.1.0&tabs=powershell&pivots=windows-psgallery#update-the-az-powershell-module

My guess is that you might not have authenticated to the PowerShell SDK prior to running the Deployment script. Have you followed the steps [here](https://azure.github.io/ipam/#/deployment/README?id=authenticate-to-azure-powershell) first?

If that doesn't work, can you share the output of the following PowerShell command:

```
(Get-AzContext).Environment.Name
```

Knowing this output will be extremely helpful in further diagnosing your issue.

Give that a whirl and let me know how things go!

**Reply by veerabose** [reply]
Hi DCMattty,

Thank you for your response and time. I tried again, I'm getting error " "#requires" statements of the script are missing: Az." this was reason I modified the version of Az version in deploy.ps1. 

Installed Az module details from my laptop
11.1.0               Az                                  PSGallery            Microsoft Azure PowerShell - Cmdlets to manage resources in Azure. This module is compat…

PS C:\Users\Admin\ipam\deploy> .\deploy.ps1 -location "eastus"
WARNING: Failed to get Function App Stack definitions from ARM API. Please open an issue at https://github.com/Azure/azure-powershell/issues with the following title: [Az.Functions] Failed to get Function App Stack definitions from ARM API.
WARNING: Failed to get Function App Stack definitions from ARM API. Please open an issue at https://github.com/Azure/azure-powershell/issues with the following title: [Az.Functions] Failed to get Function App Stack definitions from ARM API.
WARNING: Failed to get Function App Stack definitions from ARM API. Please open an issue at https://github.com/Azure/azure-powershell/issues with the following title: [Az.Functions] Failed to get Function App Stack definitions from ARM API.
WARNING: Failed to get Function App Stack definitions from ARM API. Please open an issue at https://github.com/Azure/azure-powershell/issues with the following title: [Az.Functions] Failed to get Function App Stack definitions from ARM API.
.\deploy.ps1: The script 'deploy.ps1' cannot be run because the following modules that are specified by the "#requires" statements of the script are missing: Az.
PS C:\Users\Admin\ipam\deploy> (Get-AzContext).Environment.Name
AzureCloud
PS C:\Users\Admin\ipam\deploy>

Thanks
Veera.

**Comment by DCMattyG** [comment]
Thanks for the feedback @veerabose!

I believe you likely need to reinstall the Azure PowerShell SDK, and I'd recommend using [this](https://learn.microsoft.com/en-us/powershell/azure/install-azps-windows?view=azps-11.1.0&tabs=powershell&pivots=windows-psgallery#installation) guide to do so. It's important to completely close and restart your PowerShell console too after the installation in case there are some files in use that cannot be updated live.

You may also want to review the [prerequisites](https://azure.github.io/ipam/#/deployment/README?id=prerequisites) section of our documentation. In addition to the Azure PowerShell SDK we require the Bicep CLI to be available which can be downloaded [here](https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/install#windows).

I hope all of this helps. The output of the command you shared above is indeed correct, so I believe this is just a case of making sure the tooling is properly installed and configure on your machine, then you should be good o install Azure IPAM.

**Comment by veerabose** [comment]
Hi DCMatty,

Thank you for your response and time. After following the instruction, I'm able to execute the deploy.ps1. however got below error.

(Code: InvalidTemplateDeployment)  - Validation failed for a resource. Check 'Error.Details[0]' for more information. (Code: ValidationForResourceFailed)    -
     | This region has quota of 0 PremiumV3 instances for your subscription. Try selecting different region or SKU. (Code:SubscriptionIsOverQuotaForSku)   

since im using free tire subscription. I modified the appService BICEP file to use standard appservice plan SKU. 
sku: {
    name: 'S1'
    size: 'S1'
    tier: 'Standard'
    capacity: 1

and executed the deploy.ps1 got error ""message": "Category 'AppServiceAntivirusScanAuditLogs' is not supported."

I modified bicep file  "category: 'AppServiceAntivirusScanAuditLogs'  enabled: false", but when I execute, deployment got stuck without creating appserviceplan for long time.

Could you please guide to complete the deployment using standard appserviceplan SKU.

Thanks
Veera.

**Reply by veerabose** [reply]
Hi DCMatty,

Good day, could you please guide me on how to provision the IPAM deployment using App service plan with standard SKU, if its possible, could you please share the bicep script for appService.bicep with standard SKU.

Thanks & Regards,
Veera.

**Comment by veerabose** [comment]
HI DCMatty,

I tried with different subscription now, and I'm getting below error while running the "./deploy.ps1 -Location "westus" -NamePrefix "devipam""

INFO: Deploying IPAM bicep templates
PS>TerminatingError(New-AzDeployment): "The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: 10:31:25 AM - Error: Code=InvalidTemplateDeployment; Message=The template deployment failed with multiple errors. Please see details for more information.
"
>> TerminatingError(New-AzDeployment): "The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: 10:31:25 AM - Error: Code=InvalidTemplateDeployment; Message=The template deployment failed with multiple errors. Please see details for more information.
"
>> TerminatingError(New-AzDeployment): "The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: 10:31:25 AM - Error: Code=InvalidTemplateDeployment; Message=The template deployment failed with multiple errors. Please see details for more information.
"
ERROR: Unable to deploy Azure IPAM solution due to an exception, see logs for detailed information!
Run Log: ..\logs\deploy_20231228103056AM.log
Error Log: ..\logs\error_20231228103056AM.log

Could you please help me to complete the IMAP deployment,

Thanks
Veera.

**Reply by DCMattyG** [reply]
Hi @veerabose, can you please provide the log files or the contents of them so we can determine the cause of the deployment failure?

**Reply by DCMattyG** [reply]
If the log files don't contain any useful details, you can re-run the deployment with the `-Debug` parameter to gather more verbose logs.
